### DISCUSSION TITLE
Why am I getting the 'Wall clock duration reached' Warning?

### DISCUSSION THREAD

**Discussion by alexbriannaughton** [supabase/supabase#17603] [Questions] [answered]
A couple of days ago, I started getting this warning with all of my edge function invocations. The function will execute correctly, and 5 minutes later I get this warning with a log of the CPU time used, which on average for all of my functions is less than 50ms. This warning even comes when I deploy a new Hello World function.

I'll note that I'm currently using the free tier. Also, around the time I started to get these warnings, I clicked to agree to the new organization based billing, but this detail might be a red herring : ) I've tried creating edge functions within new projects and even a new organization, and they all give me this warning.

**Comment by laktek** [comment, marked as answer]
@alexbriannaughton Do you still receive the 'Wall clock duration reached'? We previously had an incorrect behavior where the warning was emitted when it shouldn't. We've deployed a fix for it.

**Reply by SamLeatherdale** [reply]
This is still occuring as of April 2024. If I had to guess, even though the request has been responded to, isn't the `Deno.serve` instance still running and listening for requests, and will continue to do so indefinitely? 

It makes me think once a single request has been served, one should manually terminate the Deno server, but this is not alluded to in any of the Supabase docs.

**Comment by alexbriannaughton** [comment]
@laktek I do not--thanks for checking in!

I noticed the warning has now been replaced by a new INFO log which says 'isolate shutdown'. It is also triggered by a WallClockTimeLimit event, about five minutes after all of my edge function executions. Is this the intended behavior?

**Comment by dixhuit** [comment]
New Supabase user here. I still see `wall clock duration reached` after running an edge function locally for a few minutes. Is this expected? The function is basically a mish-mash of the cors & resend function examples - nothing crazy:

```ts
import { corsHeaders } from "../_shared/cors.ts"

const RESEND_API_KEY = Deno.env.get("RESEND_API_KEY")

Deno.serve(async (req) => {
  // This is needed if you're planning to invoke your function from a browser.
  if (req.method === "OPTIONS") {
    return new Response("ok", { headers: corsHeaders })
  }

  try {
    const { from, to, subject, html } = await req.json()
    const res = await fetch("https://api.resend.com/emails", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${RESEND_API_KEY}`,
      },
      body: JSON.stringify({
        from,
        to,
        subject,
        html,
      }),
    })

    const data = await res.json()

    return new Response(JSON.stringify(data), {
      headers: { ...corsHeaders, "Content-Type": "application/json" }, // Be sure to add CORS headers here too
      status: 200,
    })
  } catch (error) {
    return new Response(JSON.stringify({ error: error.message }), {
      headers: { ...corsHeaders, "Content-Type": "application/json" }, // and here
      status: 400,
    })
  }
})
```

**Reply by nyannyacha** [reply]
Hi! @dixhuit 😁
Yeah, that wall clock warning can be expected because that is hard-coded to be printed out when the worker has been terminated.

**Reply by dixhuit** [reply]
Thank you :)

**Reply by ajayvignesh01** [reply]
Wouldn't it be better to have a different message? Like the current one makes me think the function timeout out even though it successfully completed.

**Reply by sinameraji** [reply]
yea damn i spent hours thinking it's an error :D

**Comment by rookledookle** [comment]
This is still happening for me

**Reply by rookledookle** [reply]
@laktek sorry for tagging but could you look into it?

**Reply by IvoRacious** [reply]
Same problem for me. Has the bug really been solved?

**Reply by alexbriannaughton** [reply]
Luckily, it went away for me. I'm not sure it had anything to do with any action on my part. Now, I get a log that says 'shutdown' about 5 minutes after every execution, which is a lot better than getting the warning every time. I can't remember if the 'shutdown' log was there back when I was getting the warning.
